File: backend/tests/scoring.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
  scoreToGrade,
  personalizeScore,
  IngredientFlag,
  UserPriority,
} from '../../shared/scoring';

test('scoreToGrade boundaries match documented thresholds', () => {
  assert.equal(scoreToGrade(100), 'A');
  assert.equal(scoreToGrade(80), 'A');
  assert.equal(scoreToGrade(79), 'B');
  assert.equal(scoreToGrade(65), 'B');
  assert.equal(scoreToGrade(64), 'C');
  assert.equal(scoreToGrade(50), 'C');
  assert.equal(scoreToGrade(49), 'D');
  assert.equal(scoreToGrade(35), 'D');
  assert.equal(scoreToGrade(34), 'F');
  assert.equal(scoreToGrade(0), 'F');
});

function flag(overrides: Partial<IngredientFlag> = {}): IngredientFlag {
  return {
    ingredient: 'canola oil',
    severity: 2,
    category: 'SEED OIL',
    reason: 'high omega-6',
    priorities: ['seed_oils'],
    ...overrides,
  };
}

test('personalizeScore returns baseScore when user has no priorities', () => {
  const score = personalizeScore(80, [flag()], []);
  assert.equal(score, 80);
});

test('personalizeScore penalizes matching priorities', () => {
  const prios: UserPriority[] = ['seed_oils'];
  const score = personalizeScore(80, [flag({ severity: 2, priorities: ['seed_oils'] })], prios);
  // penalty = severity(2) * 5 * overlap(1) = 10
  assert.equal(score, 70);
});

test('personalizeScore ignores non-matching priorities', () => {
  const prios: UserPriority[] = ['vegan'];
  const score = personalizeScore(80, [flag({ priorities: ['seed_oils'] })], prios);
  assert.equal(score, 80);
});

test('personalizeScore floors at 0 and caps at 100', () => {
  const prios: UserPriority[] = ['seed_oils'];
  const many = Array.from({ length: 10 }, () =>
    flag({ severity: 3, priorities: ['seed_oils'] })
  );
  const score = personalizeScore(50, many, prios);
  assert.equal(score, 0);
});

test('personalizeScore handles multi-priority overlap correctly', () => {
  const prios: UserPriority[] = ['paraben_free', 'endocrine_disruptors'];
  const f = flag({
    ingredient: 'butylparaben',
    severity: 3,
    priorities: ['paraben_free', 'endocrine_disruptors'],
  });
  // overlap = 2, severity = 3 → penalty = 3 * 5 * 2 = 30
  const score = personalizeScore(90, [f], prios);
  assert.equal(score, 60);
});

File: android/app/src/test/java/app/revelio/domain/ScoringTest.kt
package app.revelio.domain

import app.revelio.domain.scoring.gradeFromScore
import app.revelio.domain.scoring.personalizeScore
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Ports the behavior guarantees from `shared/scoring.ts`:
 *   - No priorities → score unchanged.
 *   - Priorities match a flag → penalty = severity * 5 * overlap.
 *   - Score clamps to [0, 100].
 *   - Grade thresholds: A≥80, B≥65, C≥50, D≥35, F<35.
 */
class ScoringTest {

    @Test
    fun `no priorities returns base score`() {
        val flags = listOf(
            IngredientFlag(
                ingredient = "palm oil",
                severity = 2,
                category = "seed oil",
                reason = "",
                priorities = listOf("seed_oils"),
            )
        )
        assertEquals(80, personalizeScore(80, flags, emptyList()))
    }

    @Test
    fun `single overlapping priority subtracts severity times five`() {
        val flags = listOf(
            IngredientFlag(
                ingredient = "soybean oil",
                severity = 3,
                category = "seed oil",
                reason = "",
                priorities = listOf("seed_oils"),
            )
        )
        // 100 - (3 * 5 * 1) = 85
        assertEquals(85, personalizeScore(100, flags, listOf(UserPriority.SEED_OILS)))
    }

    @Test
    fun `multi-priority overlap multiplies penalty`() {
        val flags = listOf(
            IngredientFlag(
                ingredient = "red-40",
                severity = 2,
                category = "artificial dye",
                reason = "",
                priorities = listOf("artificial_additives", "endocrine_disruptors"),
            )
        )
        // overlap = 2, severity = 2 → penalty 2*5*2 = 20.  70 - 20 = 50.
        val score = personalizeScore(
            70,
            flags,
            listOf(UserPriority.ARTIFICIAL_ADDITIVES, UserPriority.ENDOCRINE_DISRUPTORS),
        )
        assertEquals(50, score)
    }

    @Test
    fun `score clamps at zero`() {
        val flags = List(10) {
            IngredientFlag(
                ingredient = "x",
                severity = 3,
                category = "test",
                reason = "",
                priorities = listOf("seed_oils"),
            )
        }
        assertEquals(0, personalizeScore(20, flags, listOf(UserPriority.SEED_OILS)))
    }

    @Test
    fun `score clamps at one hundred`() {
        // No flags → no penalty, so a 100 base stays at 100 even with priorities.
        assertEquals(100, personalizeScore(100, emptyList(), listOf(UserPriority.VEGAN)))
    }

    @Test
    fun `grade boundaries match ts source`() {
        assertEquals(Grade.A, gradeFromScore(80))
        assertEquals(Grade.B, gradeFromScore(65))
        assertEquals(Grade.C, gradeFromScore(50))
        assertEquals(Grade.D, gradeFromScore(35))
        assertEquals(Grade.F, gradeFromScore(34))
        assertEquals(Grade.A, gradeFromScore(100))
        assertEquals(Grade.F, gradeFromScore(0))
    }

    @Test
    fun `non-matching priorities do not penalize`() {
        val flags = listOf(
            IngredientFlag(
                ingredient = "sls",
                severity = 2,
                category = "sulfate",
                reason = "",
                priorities = listOf("sulfate_free"),
            )
        )
        // User cares about gluten, not sulfates — no penalty.
        assertEquals(60, personalizeScore(60, flags, listOf(UserPriority.GLUTEN_FREE)))
    }
}
